internal/wasm/module.go

```diff
@@ -170,6 +173,274 @@ func (m *Module) TypeOfFunction(funcIdx Index) *FunctionType {
+		return err
+	}
+
+	functions, globals, memories, tables := m.allDeclarations()
+
+	// The wazero specific limitation described at RATIONALE.md.
+	const maximumGlobals = 1 << 27
+	if err := m.validateGlobals(globals, maximumGlobals); err != nil {
+		return err
+	}
+
+	if err := m.validateTables(tables, globals); err != nil {
+		return err
+	}
+
+	if err := m.validateMemories(memories, globals); err != nil {
+		return err
+	}
+
+	if err := m.validateFunctions(functions, globals, memories, tables); err != nil {
+		return err
+	}
+
+	if err := m.validateExports(functions, globals, memories, tables); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *Module) validateStartSection() error {
+	// Check the start function is valid.
+	// TODO: this should be verified during decode so that errors have the correct source positions
+	if m.StartSection != nil {
+		startIndex := *m.StartSection
+		ft := m.TypeOfFunction(startIndex)
+		if ft == nil { // TODO: move this check to that a module can never be decoded invalidly
+			return fmt.Errorf("start function has an invalid type")
+		}
+		if len(ft.Params) > 0 || len(ft.Results) > 0 {
+			return fmt.Errorf("start function must have an empty (nullary) signature: %s", ft.String())
+		}
+	}
+	return nil
+}
+
+func (m *Module) validateGlobals(globals []*GlobalType, maxGlobals int) error {
+	if len(globals) > maxGlobals {
+		return fmt.Errorf("too many globals in a module")
+	}
+
+	// Global's initialization constant expression can only reference the imported globals.
+	// See the note on https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#constant-expressions%E2%91%A0
+	importedGlobals := globals[:m.ImportGlobalCount()]
+	for _, g := range m.GlobalSection {
+		if err := validateConstExpression(importedGlobals, g.Init, g.Type.ValType); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *Module) validateFunctions(functions []Index, globals []*GlobalType, memories []*MemoryType, tables []*TableType) error {
+	// The wazero specific limitation described at RATIONALE.md.
+	const maximumValuesOnStack = 1 << 27
+
+	for codeIndex, typeIndex := range m.FunctionSection {
+		if typeIndex >= m.SectionElementCount(SectionIDType) {
+			return fmt.Errorf("function type index out of range")
+		} else if uint32(codeIndex) >= m.SectionElementCount(SectionIDCode) {
+			return fmt.Errorf("code index out of range")
+		}
+
+		if err := validateFunction(
+			m.TypeSection[typeIndex],
+			m.CodeSection[codeIndex].Body,
+			m.CodeSection[codeIndex].LocalTypes,
+			functions, globals, memories, tables, m.TypeSection, maximumValuesOnStack); err != nil {
+			idx := m.SectionElementCount(SectionIDFunction) - 1
+			return fmt.Errorf("invalid function (%d/%d): %v", codeIndex, idx, err)
+		}
+	}
+	return nil
+}
+
+func (m *Module) validateTables(tables []*TableType, globals []*GlobalType) error {
+	if len(tables) > 1 {
+		return fmt.Errorf("multiple tables are not supported")
```

Comment:
@codefromthecrypt fyi now this is the place we enforce at most one table (before instantiation phase)